### PR TITLE
Apple Silicon Support, Bump Go to 1.17, refactor docker/build/x86_64/Dockerfile

### PR DESCRIPTION
**Pull request by kermieisinthehouse**
#### ✨ Apple Silicon Support
The new build chain supports M1 macs, and there is a new `stash-osx-applesilicon` build. ~~`stash-osx` has been renamed `stash-osx-intel`.~~ The new minimum  MacOS version is 10.10, Mavericks, released 2014.

**MERGE NOTE: This change requires a new stashapp/compiler release, versioned as 5. This PR will not pass checks without it.** 

#### ✨ Upgrade to Go 1.17
This means a double digit performance improvement across the board, no more GO111MODULE (except for packr which has trouble with the module system), and overall shininess. 
- ✅ Tested working on bare metal Linux, testing needed for windows/mac and the CI pipeline.

#### 🆕 Refactor docker/build/x86_64/Dockerfile
This is for local builds of stash, outside of the CI pipeline. I use it often for testing, and it was SLOW. It now makes better use of multi-stage builds and puts often-changed files/folders at the end of the image build line, for much better cache use during builds. Images have been switched to leaner alpine Linux versions, and packr bumped to latest version. 
-  ✅ Tested working on latest docker. The CI docker images are unchanged except for the Go 1.17 changes.

#### ♻️ Deleted develop/x86_64
A copy of docker/prod/x86_64 and unused.

#### 📖 Touched up docs
Clarified docker/prod/x86_64/Readme.md (linked to from front-page readme) to help get future docker users squared away more quickly.
Documented the new `make docker-build`.
Updated for Go 1.17 changes

### CODE REVIEW

**Comment by InfiniteTF** [comment]
Have you tested cross-compiling with a rebuilt compiler container? The OSX compiler is a binary blob which is not up to date IIRC, and I expect that might cause issues. For stash-box I switched cross-compilation pipeline entirely to move onto one that is up to date.

**Comment by bnkai** [comment]
- As mentioned above and here #1272 the first thing to do would be to either edit  the compiler docker container so that the osx sdk is somehow updated to a version that go 1.15+ supports or change the cross compile script/pipeline
- Is there a benefit to changing from ubuntu to alpine in the `build` docker? Building a docker locally imo should end up with the same distro as the ci builds (also it is simpler to maintain  imo)
- the `develop/x86_64/Dockerfile` was meant as an example if users wanted to build the container themselves afaik. Its difference with the prod one is that it fetches the dev stash binary (`.../latest_develop`) vs the stable one ( `.../latest`) that the prod uses. It shouldn't be removed imo.

**Comment by kermieisinthehouse** [comment]
@bnkai 

- I've updated the compiler image to support new go, and extended it to also build for M1 macs. Stash on a new platform!
- `build` docker is really only useful for local development. It builds the entire codebase and creates an image out of it, and already differed from the `ci` builds that get pushed to dockerhub. On a development machine, alpine uses less memory (great if you're comparing 2-3 running instances of stash) and has a small disk footprint.
- I'd say that both `develop` and `prod` were references, as they aren't used by any code. I didn't see the point of having two semi-identical references, but I can restore it if you'd like.

**Comment by WithoutPants** [comment]
> The new build chain supports M1 macs, and there is a new `stash-osx-applesilicon` build. `stash-osx` has been renamed `stash-osx-intel`.

The existing osx build name should remain `stash-osx` to not impact on existing osx users imo.

**Comment by kermieisinthehouse** [comment]
I'm investigating build a fat binary that supports both types. In the meantime, I seem to have broken this build (complaining of missing libs) but I did manage to produce a binary in some earlier step of development so I'll have it fixed soon

**Comment by kermieisinthehouse** [comment]
The new compiler image is now fully working on all architectures. I have reverted the name of the old macos binary to `stash-osx`, however I have kept the new makefile directive `cross-compile-osx-intel` to reduce ambiguity.

To test:
```
cd docker/compiler
make build
docker run -it -v /home/user/build/stash:/go/stash stashapp/compiler:latest /bin/bash
cd stash
make cross-compile-all
# watch your dist folder fill up with all sorts of nifties, including stash-osx-applesilicon
```

**Comment by kermieisinthehouse** [comment]
Testing the binaries is welcomed and needed.

You can find builds for Windows, MacOS-intel and MacOS-apple at https://kermie.isinthe.house/stash-go-1.17-experimental/ (hosted at least until pr is merged) or build them yourself with the steps above.

**Comment by kermieisinthehouse** [comment]
I ended up renting an M1 Mac instance and tested this PR. The MacOS binaries for both apple silicon and x86 work perfectly, after installing ffmpeg. I tested both the downloads from the readme and homebrew, which worked fine. The automatic ffmpeg download only worked if I ran the x86 version, as is expected since the ffmpeg source doesn't make arm64 binaries.

I tested scans, auto tag, generating every kind of asset, playback, scraping performers. Any Mac platform-specific issues should have come up in this testing. I've also been running this changeset in Linux for a few weeks, no issues found.

File: Makefile [hunk 1]
```diff
@@ -44,14 +44,15 @@ endif
...
+	go build $(OUTPUT) -mod=vendor -v -tags "sqlite_omit_load_extension osusergo netgo" $(GO_BUILD_FLAGS) -ldflags "$(LDFLAGS) $(EXTRA_LDFLAGS)"
 
 # strips debug symbols from the release build
-# consider -trimpath in go build if we move to go 1.13+
 build-release: EXTRA_LDFLAGS := -s -w
+build-release: GO_BUILD_FLAGS := -trimpath
 build-release: build
 
 build-release-static: EXTRA_LDFLAGS := -extldflags=-static -s -w
+build-release: GO_BUILD_FLAGS := -trimpath
```
Comment (WithoutPants) [inline Makefile:55]: Should this be `build-release-static`?

**Review by WithoutPants — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (kermieisinthehouse) [reply Makefile]: Fixed, and proofread the rest of the file for similar mistakes.

**Comment by WithoutPants** [comment]
I'm getting the following error when running `make docker-build`:

```
Step 29/36 : RUN make packr
 ---> Running in 155515c03487
export GO111MODULE=on
packr2
runtime: mlock of signal stack failed: 12
runtime: increase the mlock limit (ulimit -l) or
runtime: update your kernel to 5.3.15+, 5.4.2+, or 5.5+
fatal error: mlock failed
runtime stack:
runtime.throw(0x8c810b, 0xc)
        /usr/local/go/src/runtime/panic.go:1112 +0x72
runtime.mlockGsignal(0xc001608300)
        /usr/local/go/src/runtime/os_linux_x86.go:72 +0x107
runtime.mpreinit(0xc001606380)
        /usr/local/go/src/runtime/os_linux.go:341 +0x78
runtime.mcommoninit(0xc001606380)
        /usr/local/go/src/runtime/proc.go:630 +0x108
runtime.allocm(0xc000035800, 0x0, 0xc00050d500)
        /usr/local/go/src/runtime/proc.go:1390 +0x14e
runtime.newm(0x0, 0xc000035800)
        /usr/local/go/src/runtime/proc.go:1704 +^[[0m^[[91m0x39
runtime.startm(0xc000035800, 0xc6a700)
        /usr/local/go/src/runtime/proc.go:1869 +0x12a
runtime.handoffp(0xc000035800)
        /usr/local/go/src/runtime/proc.go:1896 +0x52
runtime.retake(0x685d20e2ba2, ^[[0m^[[91m0x0)
        /usr/local/go/src/runtime/proc.go:4628 +0x157
runtime.sysmon()
        /usr/local/go/src/runtime/proc.go:4537 +0x138
^[[0m^[[91mruntime.mstart1()
        /usr/local/go/src/runtime/proc.go:1097 +0xc3
runtime.mstart()
        ^[[0m^[[91m/usr/local/go/src/runtime/proc.go:1062 +0x6e
<snip>
```

There's a very long stack trace following with 696 goroutines.

**Comment by WithoutPants** [comment]
Also `COMPILER_IMAGE: stashapp/compiler:4` needs to be updated to `5` in the `build.yml` file.

**Comment by kermieisinthehouse** [comment]
@WithoutPants 
Please note that that error is a bug in the Linux kernel that causes silent and random memory corruption. Earlier Go versions did not detect the bug conditions and were silently corrupted, while newer Go runtimes will detect the bug and crash early, which is what is happening.

You will need to update to a recent Linux kernel, which is not specified within docker. If you're running on a Linux host, upgrade the system. If you're using windows/Mac through docker desktop, update docker desktop, as it provides the kernel for the Linux VM that runs docker.

See https://github.com/golang/go/issues/35777

**Pull request merged**